### ISSUE TITLE
Error when using packed8 together with shortlist?

### ISSUE THREAD

**Issue by Faken93** [marian-nmt/marian-dev#722]
### Bug description
Error occurs in decoding when using packed8 model together with shortlist.

  | No   shortlist | With   shortlist
-- | -- | --
model | √ | √
Packed16   model | √ | √
Packed8avx2   model | √ | ×
Packed8avx512   model | √ | ×

### Context
* Marian version:  v1.9.0
* Environment: CPU

**Comment by emjotde** [comment]
Have you tried current master in marian-dev? @ykim362 we expect this to work currently, right?

**Comment by Faken93** [comment]
I have tried _marian-dev_ yesterday. Same error occurs. @emjotde

**Comment by Faken93** [comment]
I wonder if any special config should be set? @emjotde  Thanks!

[config] alignment: ""
[config] allow-unk: false
[config] authors: false
[config] beam-size: 1
[config] bert-class-symbol: "[CLS]"
[config] bert-mask-symbol: "[MASK]"
[config] bert-masking-fraction: 0.15
[config] bert-sep-symbol: "[SEP]"
[config] bert-train-type-embeddings: true
[config] bert-type-vocab-size: 2
[config] best-deep: false
[config] build-info: ""
[config] cite: false
[config] clip-gemm: 0
[config] cpu-threads: 2
[config] dec-cell: ssru
[config] dec-cell-base-depth: 1
[config] dec-cell-high-depth: 1
[config] dec-depth: 3
[config] devices:
[config]   - 0
[config] dim-emb: 512
[config] dim-rnn: 512
[config] dim-vocabs:
[config]   - 23000
[config]   - 22997
[config] dump-config: ""
[config] enc-cell: gru
[config] enc-cell-depth: 1
[config] enc-depth: 25
[config] enc-type: bidirectional
[config] ignore-model-config: false
[config] input:
[config]   - stdin
[config] input-types:
[config]   []
[config] interpolate-env-vars: false
[config] layer-normalization: false
[config] lemma-dim-emb: 0
[config] log: ""
[config] log-level: info
[config] log-time-zone: ""
[config] max-length: 256
[config] max-length-crop: true
[config] max-length-factor: 3
[config] maxi-batch: 1000
[config] maxi-batch-sort: src
[config] mini-batch: 32
[config] mini-batch-words: 0
[config] models:
[config]   - /home/work/model/tf-ssru_180-pack8.bin
[config] n-best: false
[config] normalize: 1
[config] num-devices: 0
[config] optimize: true
[config] output: stdout
[config] output-sampling: false
[config] port: 8080
[config] precision:
[config]   - float32
[config] quiet: false
[config] quiet-translation: true
[config] relative-paths: false
[config] right-left: false
[config] seed: 0
[config] shortlist:
[config]   - lex.s2t
[config]   - 100
[config]   - 75
[config] skip: false
[config] skip-cost: false
[config] tied-embeddings: false
[config] tied-embeddings-all: false
[config] tied-embeddings-src: false
[config] transformer-aan-activation: swish
[config] transformer-aan-depth: 2
[config] transformer-aan-nogate: false
[config] transformer-decoder-autoreg: rnn
[config] transformer-depth-scaling: false
[config] transformer-dim-aan: 2048
[config] transformer-dim-ffn: 2048
[config] transformer-ffn-activation: relu
[config] transformer-ffn-depth: 2
[config] transformer-guided-alignment-layer: last
[config] transformer-heads: 8
[config] transformer-no-projection: false
[config] transformer-postprocess: da
[config] transformer-postprocess-emb: d
[config] transformer-preprocess: n
[config] transformer-tied-layers:
[config]   []
[config] transformer-train-position-embeddings: false
[config] type: transformer
[config] vocabs:
[config]   - /home/work/model/sr.vocab.yml
[config]   - /home/work/model/en.vocab.yml
[config] weights:
[config]   []
[config] word-penalty: 0
[config] word-scores: false
[config] workspace: 2000

**Comment by XapaJIaMnu** [comment]
@Faken93 can you post the error log?

**Comment by Faken93** [comment]
[Error log]
Error: No hypotheses in n-best list??
@XapaJIaMnu

**Comment by Faken93** [comment]
In src\layers\generic.cpp, 
_Logits Output::applyAsLogits(Expr input) {
...
   if (shortlist_ && !cachedShortWt_) { // shortlisted versions of parameters are cached within one batch, then clear()ed
        cachedShortWt_ = index_select(Wt_, isLegacyUntransposedW ? -1 : 0, shortlist_->indices());
        cachedShortb_  = index_select(b_ ,                             -1, shortlist_->indices());
      }
...
}_

Here, when there is shortlist, the _Wt__ will be converted to _cachedShortWt__ by function _index_select_. But, if the type of Wt_ is packed8***, I wonder whether function _index_select_ is as expected. Thanks!

_fbgemmPacked8Pack{
...
  memcpy(packedBuf + (packsize - n * (sizeof(float) + sizeof(int32_t) + sizeof(int32_t))), quantScaleB, n * sizeof(float));
  memcpy(packedBuf + (packsize - n * (sizeof(int32_t) + sizeof(int32_t))), quantZeropointB, n * sizeof(int32_t));
  memcpy(packedBuf + (packsize - n * sizeof(int32_t)), colOffsets, n * sizeof(int32_t));
}_

**Comment by ykim362** [comment]
@emjotde Yes, this should work. @Faken93 With shortlist, Wt_ should stay as fp32. The current code doesn't pack embedding matrix. I assumes that you are converting the model offline into packed format. Then, could you try to turn off `optimize` flag in your decoding command? `[config] optimize: true`

**Comment by XapaJIaMnu** [comment]
@Faken93, I just reproduced your problem in a GPU environment. 2080tis fail, Pascal Titan X works. What CPU are you using? Is it VNNI capable?

@emjotde `ENABLE_CUBLAS_TENSOR_OP_MATH_FP32=0` did not help. Are we running into some obscure floating point bugs?

**Comment by XapaJIaMnu** [comment]
@Faken93 , could you post your CPU. Is it VNNI.

**Comment by ykim362** [comment]
@XapaJIaMnu I am not 100% sure, but I suspect about the issue of mixed int16 and int8. `optimize` enables int16 gemm where the weights are not pre-packed to int8. Most of the performance related weights are already pre-packed to int8, so I don't think we can get additional benefits from int16 (maybe, it could slow down). When we have `optimize` with int8, we may get some more performance benefits.

**Comment by XapaJIaMnu** [comment]
@ykim362 #595 does exactly that. 
In master, `--optimize` doesn't do the output layer (i think) so there would be no point form it.

**Comment by ykim362** [comment]
@XapaJIaMnu Yes, that's what I was saying.

**Comment by Faken93** [comment]
1. Yeah, @ykim362 ,you are right. All models of packed format are converted offline.  
2. As you said, I keep _Wt__ as fp32, and it works successfully. But, I found that, _Wt_(fp32) with shortlist_ is slower than _Wt_(packed8) without shorlist_. On the one hand, I'm so surprised at the packed8 performance. On the other hand, **I wonder whether the result is as expected.**
3. I have tried --optimize: false/true, there is no difference.

@emjotde @XapaJIaMnu @ykim362 Thanks!

**Comment by ykim362** [comment]
@Faken93 Did you modify the marian code to pack the embedding matrix into int8? If that's the case, this is expected, because packed format is a tiled transposed special matrix format. Shortlist doesn't work directly with the packed format. As @XapaJIaMnu mentioned, we have a PR which enables int8 quantization of the embedding matrix as well. #595

**Comment by Faken93** [comment]
I didn't modify the marian conv code and I think _Wt__ is not an embedding matrix. @ykim362 

The _Wt__ tensor name is _decoder_ff_logit_out_W_ in my model. It appears in the final feed-forward layer of transformer.

src\model\tranformer.h
Ptr<DecoderState> step(Ptr<DecoderState> state) {
    ...
   
    // final feed-forward layer (output)
    if(shortlist_)
      output_->setShortlist(shortlist_);
    auto logits = output_->applyAsLogits(decoderContext); 
   
    ...
  }

src\layers\generic.cpp,
Logits Output::applyAsLogits(Expr input) {
  ...
  if (shortlist && !cachedShortWt_) { 
     cachedShortWt_ = index_select(Wt_, isLegacyUntransposedW ? -1 : 0, shortlist_->indices());
     cachedShortb_ = index_select(b_ , -1, shortlist_->indices());
   }
  ...
}

**Comment by Faken93** [comment]
I also tried to use #595 , but I cannot find option --optimize8 in https://github.com/marian-nmt/marian-dev/blob/intgemm_reintegrated/src/common/config_parser.cpp

**Comment by ykim362** [comment]
@Faken93 Ah, I see what's going on now. In many cases, `[config] tied-embeddings-all` is set to true. I always use it true. In that case, the output matrix `Wt_` is `Wemb`. marian-conv doesn't pack `Wemb`. That's why I wondered how `Wemb` - `Wt_` is packed. To make the shortlist working, _decoder_ff_logit_out_W_ should not be packed. index_select doesn't work with packed format at the moment.

**Comment by Faken93** [comment]
Yeah, thanks for your reply, it is very clear. @ykim362

**Comment by XapaJIaMnu** [comment]
> I also tried to use #595 , but I cannot find option --optimize8 in https://github.com/marian-nmt/marian-dev/blob/intgemm_reintegrated/src/common/config_parser.cpp

@Faken93 the option was renamed to `--int16`. There's a new one that does `--int8` as well. Those can be used in conjunction with fbgemm packed models (the ones that you are using.) and will pack the embeddings on the fly. This should be faster than master

**Comment by Faken93** [comment]
Great, I will have a try! Thanks! @XapaJIaMnu
